### PR TITLE
perf(goal_planner): reduce processing time 

### PR DESCRIPTION
**Pull request by kosuke55**
## Description

reduce proccesing time

- Filter objects in advance.
  - (the most effective part in this PR)
- Calculate path curvature in advance
- rough collision check 
  - https://github.com/autowarefoundation/autoware.universe/pull/8193 
  - If it's safe at min_distance, it's safe
  - If max_distance is unsafe, it's unsafe.
  - Otherwise, use footprint to make the decision,

when `needPathUpdate()`  (once per 1s)
before:

60-70ms 
image

image

after:
20-30ms 

object is not related to the time so much.

image

image

not when `needPathUpdate()`  
after

~10 ms

image

## Related links

**Parent Issue:**

- Link

**Private Links:**

- https://tier4.atlassian.net/browse/RT1-6970

## How was this PR tested?

psim
2024/07/25 https://evaluation.tier4.jp/evaluation/reports/04c830e5-290e-5a36-8702-526a5fba407d/?project_id=prd_jt

3589/3590

## Notes for reviewers

None.

## Interface changes

None.

## Effects on system behavior

None.

### CODE REVIEW

**Comment by kosuke55** [comment]
include https://github.com/autowarefoundation/autoware.universe/pull/8194

File: planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/include/autoware/behavior_path_goal_planner_module/pull_over_planner_base.hpp [hunk 1]
```diff
@@ -52,41 +52,28 @@ struct PullOverPath
...
+    updatePathData();
   }
 
-  // path from the pull over start pose to the end pose
-  PathWithLaneId getParkingPath() const
-  {
-    const PathWithLaneId full_path = getFullPath();
-    const size_t start_idx =
-      autoware::motion_utils::findNearestIndex(full_path.points, start_pose.position);
+  PathWithLaneId getFullPath() { return full_path; }
```
Comment (soblin) [inline planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/include/autoware/behavior_path_goal_planner_module/pull_over_planner_base.hpp:74]: We should return const& reference

File: planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/include/autoware/behavior_path_goal_planner_module/pull_over_planner_base.hpp [hunk 2]
```diff
@@ -52,41 +52,28 @@ struct PullOverPath
...
 
-    PathWithLaneId parking_path{};
-    std::copy(
-      full_path.points.begin() + start_idx, full_path.points.end(),
-      std::back_inserter(parking_path.points));
-
-    return parking_path;
-  }
+  PathWithLaneId getParkingPath() const { return parking_path; }
```
Comment (soblin) [inline planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/include/autoware/behavior_path_goal_planner_module/pull_over_planner_base.hpp:76]: same as asbove

File: planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp [hunk 3]
```diff
@@ -1048,19 +1079,22 @@ std::optional<std::pair<PullOverPath, GoalCandidate>> GoalPlannerModule::selectP
       continue;
     }
 
-    const auto resampled_path = utils::resamplePathWithSpline(pull_over_path.getParkingPath(), 0.5);
+    // const auto resampled_path = utils::resamplePathWithSpline(pull_over_path.getParkingPath(),
+    // 0.5);
+    const PathWithLaneId parking_path = pull_over_path.getParkingPath();
```
Comment (soblin) [inline planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp:1084]: and bind by const reference &

File: planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp [hunk 4]
```diff
@@ -911,11 +938,13 @@ std::vector<PullOverPath> GoalPlannerModule::sortPullOverPathCandidatesByGoalPri
         return 0.0;
       }
       // check path collision margin
-      const auto resampled_path =
-        utils::resamplePathWithSpline(pull_over_path.getParkingPath(), 0.5);
+      // const auto resampled_path =
+      //   utils::resamplePathWithSpline(pull_over_path.getParkingPath(), 0.5);
```
Comment (kosuke55) [inline planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp:942]:
```suggestion
```

**Review by soblin — APPROVED** [review]

File: planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/include/autoware/behavior_path_goal_planner_module/pull_over_planner_base.hpp (hunk 1 above)
Comment (kosuke55) [reply planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/include/autoware/behavior_path_goal_planner_module/pull_over_planner_base.hpp]: fixed in dbd9ff70604526c6c8b8ef5b142a31c112c0a724

File: planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/include/autoware/behavior_path_goal_planner_module/pull_over_planner_base.hpp (hunk 2 above)
Comment (kosuke55) [reply planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/include/autoware/behavior_path_goal_planner_module/pull_over_planner_base.hpp]: dbd9ff70604526c6c8b8ef5b142a31c112c0a724

File: planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/include/autoware/behavior_path_goal_planner_module/pull_over_planner_base.hpp (hunk 1 above)
Comment (kosuke55) [reply planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/include/autoware/behavior_path_goal_planner_module/pull_over_planner_base.hpp]: revert in [use copy getter](https://github.com/autowarefoundation/autoware.universe/pull/8195/commits/8ce336e0a1d1dddfb2ce7a10e82b97cab7685401)

becase the returned value is used in multi threads

File: planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp (hunk 3 above)
Comment (kosuke55) [reply planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/src/goal_planner_module.cpp:1082]: revert in [use copy getter](https://github.com/autowarefoundation/autoware.universe/pull/8195/commits/8ce336e0a1d1dddfb2ce7a10e82b97cab7685401)

becase the returned value is used in multi threads

File: planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/include/autoware/behavior_path_goal_planner_module/pull_over_planner_base.hpp (hunk 2 above)
Comment (kosuke55) [reply planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/include/autoware/behavior_path_goal_planner_module/pull_over_planner_base.hpp:77]: revert in [use copy getter](https://github.com/autowarefoundation/autoware.universe/pull/8195/commits/8ce336e0a1d1dddfb2ce7a10e82b97cab7685401)

becase the returned value is used in multi threads

**Pull request merged**